cache/cache.go

```diff
@@ -13,3 +13,180 @@
+
+	cache := &Cache{
+		prefix:  prefix,
+		cfg:     cfg,
+		client:  client,
+		history: ringBuffer,
+		demux:   newDemux(ringBuffer, cfg.ResyncInterval),
+		ctx:     ctx,
+		cancel:  cancel,
+		ready:   make(chan struct{}),
+	}
+
+	go func() {
+		// initial catch up
+		if err := cache.initialLoad(); err != nil {
+			log.Printf("cache initial load failed for prefix %q: %v",
+				prefix, err)
+		}
+		close(cache.ready)
+
+		// long-running upstream watch
+		serveWatchEvents(
+			ctx, client, prefix,
+			cache.demux.broadcast,
+			cache.demux,
+			cache.history,
+			func(rev int64) { atomic.StoreInt64(&cache.latestCompactedRev, rev) },
+			cfg.InitialBackoff,
+			cfg.MaxBackoff,
+		)
+	}()
+
+	return cache, nil
+}
+
+// initialLoad ranges over the prefix and seeds the ring.
+func (c *Cache) initialLoad() error {
+	opts := []clientv3.OpOption{clientv3.WithPrefix()}
+	resp, err := c.client.Get(c.ctx, c.prefix, opts...)
+	if err != nil {
+		return err
+	}
+	for _, kv := range resp.Kvs {
+		event := &clientv3.Event{
+			Type: clientv3.EventTypePut,
+			Kv:   kv,
+		}
+		c.history.Append(event)
+	}
+	return nil
+}
+
+func (c *Cache) Watch(ctx context.Context, key string, opts ...clientv3.OpOption) clientv3.WatchChan {
+	select {
+	case <-c.ready:
+	case <-ctx.Done():
+		emptyWatchChan := make(chan clientv3.WatchResponse)
+		close(emptyWatchChan)
+		return emptyWatchChan
+	}
+
+	if !strings.HasPrefix(key, c.prefix) {
+		// invalid key
+		emptyWatchChan := make(chan clientv3.WatchResponse)
+		close(emptyWatchChan)
+		return emptyWatchChan
+	}
+
+	op := clientv3.OpGet(key, opts...)
+	requestedRev := op.Rev()
+
+	responseChan := make(chan clientv3.WatchResponse, 1)
+
+	// compaction / gap checks
+	if requestedRev > 0 && requestedRev < c.history.OldestRevision() {
+		responseChan <- clientv3.WatchResponse{CompactRevision: c.history.OldestRevision()}
+		close(responseChan)
+		return responseChan
+	}
+
+	stream, err := c.newWatchStream(ctx, key, requestedRev)
+	if err != nil {
+		responseChan <- clientv3.WatchResponse{Canceled: true}
+		close(responseChan)
+		return responseChan
+	}
+
+	go func() {
+		defer close(responseChan)
+		for {
+			select {
+			case ev, ok := <-stream.Ch:
+				if !ok {
+					return
+				}
+				responseChan <- clientv3.WatchResponse{Events: []*clientv3.Event{ev}}
+			case <-ctx.Done():
+				return
+			}
+		}
+	}()
+	return responseChan
+}
+
+func (c *Cache) newWatchStream(ctx context.Context, key string, startRev int64) (*Stream, error) {
+	pred := func(k []byte) bool { return strings.HasPrefix(string(k), key) }
+
+	replay, last := c.history.GetSince(startRev, pred)
+	w := newWatcher(c.cfg.PerWatcherBufferSize, pred, last)
+
+	for _, ev := range replay {
+		select {
+		case w.eventChan <- ev:
+		case <-ctx.Done():
+			w.stop()
+			return nil, ctx.Err()
+		}
+	}
+
+	c.demux.add(w)
+	w.autoRemoveOnCancel(ctx, c.demux)
+
+	return &Stream{Ch: w.eventChan}, nil
+}
+
+// Close shuts down the shard.
+func (c *Cache) Close() { c.cancel() }
```

Comment:
We aren't using this anywhere other than tests. We should call this in the constructor using a `defer` after starting the caching. 

Also, since this is just calling `c.cancel()`, we can call `defer cancel()` directly in the constructor and not have to keep `cancel` as part of the struct.